hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RocksDatabase.java

```diff
@@ -0,0 +1,490 @@
+        return d.getClass() + ":" + bytes2String(d.getName());
+      }
+    });
+  }
+
+  private static void close(Map<String, ColumnFamily> columnFamilies,
+      RocksDB db, List<ColumnFamilyDescriptor> descriptors,
+      WriteOptions writeOptions, DBOptions dbOptions) {
+    if (columnFamilies != null) {
+      for (ColumnFamily f : columnFamilies.values()) {
+        runWithTryCatch(() -> f.getHandle().close(), f);
+      }
+    }
+
+    if (db != null) {
+      runWithTryCatch(db::close, "db");
+    }
+
+    if (descriptors != null) {
+      descriptors.forEach(RocksDatabase::close);
+    }
+
+    if (writeOptions != null) {
+      runWithTryCatch(writeOptions::close, "writeOptions");
+    }
+    if (dbOptions != null) {
+      runWithTryCatch(dbOptions::close, "dbOptions");
+    }
+  }
+
+  private static void runWithTryCatch(Runnable runnable, Object name) {
+    try {
+      runnable.run();
+    } catch (Throwable t) {
+      LOG.error("Failed to close " + name, t);
+    }
+  }
+
+  static Predicate<TableConfig> extraColumnFamily(Set<TableConfig> families) {
+    return f -> {
+      if (families.contains(f)) {
+        return false;
+      }
+      LOG.info("Found an extra column family in existing DB: {}", f);
+      return true;
+    };
+  }
+
+  public boolean isClosed() {
+    return isClosed.get();
+  }
+
+  /**
+   * Represents a checkpoint of the db.
+   *
+   * @see Checkpoint
+   */
+  final class RocksCheckpoint {
+    private final Checkpoint checkpoint;
+
+    private RocksCheckpoint() {
+      this.checkpoint = Checkpoint.create(db);
+    }
+
+    public void createCheckpoint(Path path) throws IOException {
+      try {
+        checkpoint.createCheckpoint(path.toString());
+      } catch (RocksDBException e) {
+        closeOnError(e);
+        throw toIOException("createCheckpoint " + path, e);
+      }
+    }
+
+    public long getLatestSequenceNumber() {
+      return RocksDatabase.this.getLatestSequenceNumber();
+    }
+  }
+
+  /**
+   * Represents a column family of the db.
+   *
+   * @see ColumnFamilyHandle
+   */
+  public static final class ColumnFamily {
+    private final byte[] nameBytes;
+    private final String name;
+    private final ColumnFamilyHandle handle;
+
+    public ColumnFamily(ColumnFamilyHandle handle) throws RocksDBException {
+      this.nameBytes = handle.getName();
+      this.name = bytes2String(nameBytes);
+      this.handle = handle;
+      LOG.debug("new ColumnFamily for {}", name);
+    }
+
+    public String getName() {
+      return name;
+    }
+
+    public String getName(StringCodec codec) {
+      return codec.fromPersistedFormat(nameBytes);
+    }
+
+    private ColumnFamilyHandle getHandle() {
+      return handle;
+    }
+
+    public int getID() {
+      return getHandle().getID();
+    }
+
+    public void batchDelete(WriteBatch writeBatch, byte[] key)
+        throws IOException {
+      try {
+        writeBatch.delete(getHandle(), key);
```

Comment:
WriteBatch needs to be closed if unused. See PR #3400

Reply:
@jojochuang I will be filling additional PRs to address remaining leaks. This PR is to get the RocksDatabase.java abstraction in and address the cache handling for it.